picocli-annotation-processing-tests/src/test/java/picocli/annotation/processing/tests/AbstractCommandSpecProcessorTest.java

```diff
@@ -225,9 +226,49 @@ public void testInvalidSplitOnSinglValueOptionOrPositional() {
         validateErrorMessages(compilation, expected, optional);
     }
 
+    @Test
+    public void testInvalidFinalOptionsAndParameters() {
+        CommandSpec2YamlProcessor processor = new CommandSpec2YamlProcessor();
+        Compilation compilation =
+            javac()
+                .withProcessors(processor)
+                .compile(JavaFileObjects.forResource(
+                    "picocli/examples/validation/InvalidFinal.java"));
+
+        assertThat(compilation).failed();
+
+        // For every primitive type + String type, the InvalidFinal class defines
+        // an invalid combination of using a final field with a declared value, for each of those types.
+        List<String> types = Arrays.asList(
+            "boolean",
+            "byte",
+            "short",
+            "int",
+            "long",
+            "char",
+            "float",
+            "double",
+            "string"
+        );
+
+        String errorFormat = "Constant (final) primitive and String fields like %s cannot be used as %s: compile-time constant inlining may hide new values written to it.";
+        List<String> expectedValidationErrors = new ArrayList<String>();
+        for (String type : types) {
+            String titleized = type.substring(0, 1).toUpperCase() + type.substring(1);
+            String invalidOptionField = String.format("invalid%s", titleized);
+            String invalidParamField = String.format("invalid%sParam", titleized);
+
+            expectedValidationErrors.add(String.format(errorFormat, invalidOptionField, "@Option"));
+            expectedValidationErrors.add(String.format(errorFormat, invalidParamField, "@Parameters"));
+        }
```

Comment:
Oh, I also wasn't certain about the java version where these tests get executed, but I demonstrated considerable restraint and wrote this test to only use java 6+ language features.